File: adt2keyb/sdl/GetKey.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static uint16_t GetKey_proc (void) {
  while (true) {
    SDL_Event e;

    draw_screen ();

    if (SDL_PollEvent (&e)) {
      if ((e.type == SDL_QUIT) || _force_program_quit) {
        _force_program_quit = true;
        return kESC;
      }

      // Skip all other event except key presses
      if (e.type != SDL_KEYDOWN)
        continue;
      else {
        unsigned i;

        // Skip all modifier keys
        if (e.key.keysym.sym >= SDLK_NUMLOCK)
          continue;

        // Roll through `symtab', form correct `GetKey' value
        for (i = 0; i <= SYMTABSIZE; i++)
          if (e.key.keysym.sym == symtab[i].SDLK) {
            AT2ExtKeyCode k;

            // First check with modifier keys, order: ALT, CTRL, SHIFT (as DOS does)

            // ALT
            if (keydown[SC_LALT] || keydown[SC_RALT]) {
              k = symtab[i].alt;

              if (k == ATEK_NULL) // HINT: Intersects with unknown so should be placed first
                return k;
              else if (ATEK_IsUnknown (k)) // Impossible combination
                continue;
              else if (ATEK_IsKey (k) || ATEK_IsModifier (k)) // Found
                return k;
              else
                return MK_EKEYSC (symtab[i].scan, ATEK_GetChar (k)); // Character
            }

            // CTRL
            if (keydown[SC_LCTRL] || keydown[SC_RCTRL]) {
              k = symtab[i].ctrl;

              if (k == ATEK_NULL) // HINT: Intersects with unknown so should be placed first
                return k;
              else if (ATEK_IsUnknown (k)) // Impossible combination
                continue;
              else if (ATEK_IsKey (k) || ATEK_IsModifier (k)) // Found
                return k;
              else
                return MK_EKEYSC (symtab[i].scan, ATEK_GetChar (k)); // Character
            }

            // SHIFT
            if (keydown[SC_LSHIFT] || keydown[SC_RSHIFT]) {
              k = symtab[i].shift;

              if (_capslock)
                k = symtab[i].caps_sh;
              else if (_numlock)
                k = symtab[i].num_sh;

              if (k == ATEK_NULL) // HINT: Intersects with unknown so should be placed first
                return k;
              else if (ATEK_IsUnknown (k)) // Impossible combination
                continue;
              else if (ATEK_IsKey (k) || ATEK_IsModifier (k)) // Found
                return k;
              else
                return MK_EKEYSC (symtab[i].scan, ATEK_GetChar (k)); // Character
            }

            // Normal symbol (ASCII)
            k = symtab[i].key;

            if (_capslock)
              k = symtab[i].caps;
            else if (_numlock)
              k = symtab[i].num;

            if (k == ATEK_NULL) // HINT: Intersects with unknown so should be placed first
              return k;
            else if (ATEK_IsUnknown (k)) // Impossible combination
              continue;
            else if (ATEK_IsKey (k) || ATEK_IsModifier (k)) // Found
              return k;
            else
              return MK_EKEYSC (symtab[i].scan, ATEK_GetChar (k)); // Character
          }
      }
    }
  }
}

uint16_t GetKey (void) {
  do {
    draw_screen ();
  } while (!keypressed ());

  // Filter out CTRL+TAB combo as it is handled within timer routine
  if (ctrl_tab_pressed ()) {
    draw_screen ();
    keyboard_reset_buffer ();
    return ATEK_UNKNOWN;
  } else
    return GetKey_proc ();
}

File: adt2keyb/sdl/pas/GetKey.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function GetKey: Word;

function getkey_proc: Word;

var
  event: SDL_Event;
  i,j: Integer;

begin
  Repeat
    draw_screen;
    If (SDL_PollEvent(@event) <> 0) then
      begin
        If (event.eventtype = SDL_EVENTQUIT) or _force_program_quit then
          begin
            _force_program_quit := TRUE;
            getkey_proc := kESC;
            EXIT;
          end;
        // skip all other event except key presses
        If (event.eventtype <> SDL_KEYDOWN) then CONTINUE
        else
          begin
            // skip all modifier keys
            If (event.key.keysym.sym >= SDLK_NUMLOCK) then CONTINUE;
            // roll thru symtab, form correct GetKey value
            For j := 0 to PRED(SYMTABSIZE) do
              begin
                If (event.key.keysym.sym = symtab[j*10+9]) then
                  begin // first check with modifier keys, order: ALT, CTRL, SHIFT (as DOS does)
                    { ALT }
                    If (keydown[SC_LALT] = TRUE) or (keydown[SC_RALT] = TRUE) then
                      begin
                        // impossible combination
                        If (symtab[j*10+4] = WORD_NULL) then CONTINUE;
                        If (symtab[j*10+4] > BYTE_NULL) then
                          begin
                            getkey_proc := symtab[j*10+4];
                            EXIT;
                          end;
                        getkey_proc := (symtab[j*10] SHL 8) OR symtab[j*10+4];
                        EXIT;
                      end;
                    { CTRL }
                    If (keydown[SC_LCTRL] = TRUE) or (keydown[SC_RCTRL] = TRUE) then
                      begin
                        // impossible combination
                        If (symtab[j*10+3] = WORD_NULL) then CONTINUE;
                        If (symtab[j*10+3] > BYTE_NULL) then
                          begin
                            getkey_proc := symtab[j*10+3];
                            EXIT;
                          end;
                        getkey_proc := (symtab[j*10] SHL 8) OR symtab[j*10+3];
                        EXIT;
                      end;
                    { SHIFT }
                    If (keydown[SC_LSHIFT] = TRUE) or (keydown[SC_RSHIFT] = TRUE) then
                      begin
                        i := 2; // SHIFT
                        If (_capslock = TRUE) then i := 7 // caps lock
                        else If (_numlock = TRUE) then i := 8; // num lock
                        // impossible combination
                        If (symtab[j*10+i] = WORD_NULL) then CONTINUE;
                        If (symtab[j*10+i] > BYTE_NULL) then getkey_proc := symtab[j*10+i]
                        else getkey_proc := (symtab[j*10] SHL 8) OR symtab[j*10+i];
                        EXIT;
                      end;
                    { normal ASCII }
                    i := 1;
                    If (_capslock = TRUE) then i := 6 // caps lock
                    else If (_numlock = TRUE) then i := 5; // num lock
                    // impossible combination
                    If (symtab[j*10+i] = WORD_NULL) then CONTINUE;
                    If (symtab[j*10+i] > BYTE_NULL) then getkey_proc := symtab[j*10+i]
                    else getkey_proc := (symtab[j*10] SHL 8) OR symtab[j*10+i]; // (scancode << 8) + ASCII
                    EXIT;
                  end;
              end;
          end;
      end;
  until FALSE;
end;

begin
  Repeat draw_screen until keypressed;
  // filter out CTRL+TAB combo as it is handled within timer routine
  If ctrl_tab_pressed then
    begin
      draw_screen;
      keyboard_reset_buffer;
      GetKey := WORD_NULL;
    end
  else GetKey := getkey_proc;
end;
